nautobot_ssot/integrations/slurpit/diffsync/models/models.py

```diff
@@ -0,0 +1,330 @@
+    physical_address: Optional[str]
+    latitude: Optional[float]
+    longitude: Optional[float]
+    tags: List[TagDict] = []
+    system_of_record: Annotated[str, CustomFieldAnnotation(name="system_of_record", key="system_of_record")]
+    last_synced_from_sor: Annotated[str, CustomFieldAnnotation(name="last_synced_from_sor", key="last_synced_from_sor")]
+
+
+class ManufacturerModel(ModelQuerySetMixin, NautobotModel):
+    """Data model representing a Manufacturer."""
+
+    _model = Manufacturer
+    _modelname = "manufacturer"
+    _identifiers = ("name",)
+    _attributes = ("system_of_record", "last_synced_from_sor")
+
+    name: str
+    system_of_record: Annotated[str, CustomFieldAnnotation(name="system_of_record", key="system_of_record")]
+    last_synced_from_sor: Annotated[str, CustomFieldAnnotation(name="last_synced_from_sor", key="last_synced_from_sor")]
+
+
+class DeviceTypeModel(ModelQuerySetMixin, NautobotModel):
+    """Data model representing a DeviceType."""
+
+    _model = DeviceType
+    _modelname = "device_type"
+    _identifiers = ("model", "manufacturer__name")
+    _attributes = ("tags", "system_of_record", "last_synced_from_sor")
+
+    model: str
+    manufacturer__name: str
+    tags: List[TagDict] = []
+    system_of_record: Annotated[str, CustomFieldAnnotation(name="system_of_record", key="system_of_record")]
+    last_synced_from_sor: Annotated[str, CustomFieldAnnotation(name="last_synced_from_sor", key="last_synced_from_sor")]
+
+
+class PlatformModel(ModelQuerySetMixin, NautobotModel):
+    """Data model representing a Platform."""
+
+    _model = Platform
+    _modelname = "platform"
+    _identifiers = ("name", "manufacturer__name")
+    _attributes = ("network_driver", "napalm_driver", "system_of_record", "last_synced_from_sor")
+
+    name: str
+    manufacturer__name: str
+    network_driver: str
+    napalm_driver: str
+    system_of_record: Annotated[str, CustomFieldAnnotation(name="system_of_record", key="system_of_record")]
+    last_synced_from_sor: Annotated[str, CustomFieldAnnotation(name="last_synced_from_sor", key="last_synced_from_sor")]
+
+
+class RoleModel(ModelQuerySetMixin, NautobotModel):
+    """Data model representing a Role."""
+
+    _model = Role
+    _modelname = "role"
+    _identifiers = ("name",)
+    _attributes = (
+        "content_types",
+        "color",
+        "system_of_record",
+        "last_synced_from_sor",
+    )
+
+    name: str
+    color: Optional[str]
+    content_types: List[ContentTypeDict] = []
+    system_of_record: Annotated[str, CustomFieldAnnotation(name="system_of_record", key="system_of_record")]
+    last_synced_from_sor: Annotated[str, CustomFieldAnnotation(name="last_synced_from_sor", key="last_synced_from_sor")]
+
+    @classmethod
+    def get_queryset(cls, data):
+        """Get the queryset for the Role model."""
+        return cls._model.objects.filter(name=constants.DEFAULT_DEVICE_ROLE)
+
+
+class DeviceModel(ModelQuerySetMixin, NautobotModel):
```

Comment:
I see you importing IP Addresses and assigning to Interfaces but no assigning of primary IP to the Device. Is there a reason for that? We need a primary IP assigned to Devices in order for NAPALM functionality to work and is central to most automation efforts. Typically we suggest using the management IP of the Device as primary.

Reply:
Fixed this in most recent commit. Good catch.

Reply:
I see that you added the primary_ip attribute to the Device NautobotModel. Even though that attribute is on the Device object you actually can't assign the primary IP to a Device through the Device NautobotModel. This is because for an IPAddress to be assigned as a primary IP address it must first be created and assigned to the Device's Interface. Typically with the contrib pattern we suggest putting the primary_ip assignment on the IPAddressToInterface associated NautobotModel or creating a unique model to do just the primary IP assignment. As this is currently being done will error out and cause the Device to not be created.

Reply:
It would be good to see an example of this as I could not get this to work. Looking up interface by name only could return duplicate interface names. Would you be able to do `interface__name` as well as `interface__device__name` to specifically lookup interfaces that belong to specific devices to remove duplicates?

Ideally I want to try and use the contrib `NautobotModel` instead of making custom implementations. Currently the other integration that uses this contrib diffsync model does not assign IP's to devices so no real example of how this should work. 

```
class IPAddressToInterfaceModel(NautobotModel):
    """Data model representing an IPAddress to Interface relationship."""

    _model = IPAddressToInterface
    _modelname = "ipaddress_to_interface"
    _identifiers = ("ipaddress__host", "interface__name")
    _attributes = (
        "primary_ip",
    )

    ipaddress__host: str
    interface__name: str
    primary_ip: bool
```